# HiTE Now Includes panHiTE Support!
![Version](https://img.shields.io/badge/version-0.0.1--beta-orange)

We are excited to announce a significant update to HiTE, which now includes the powerful panHiTE functionality. This release introduces the following key features:

1. **Improved LTR Detection**: The original `LTR_retriever` has been replaced with our new tool, `FiLTR`, offering improved **accuracy and completeness** in LTR detection.
2. **Introducing panHiTE**: This new workflow is designed for population genomics, facilitating TE annotation and analysis across multiple genomes. panHiTE accepts `RNA-seq` data, enabling the detection of gene expression changes associated with **specific TE insertions**, making it ideal for exploring individual-specific traits.

---

### Output Preview

- **DE_genes_from_TEs.tsv**  
This file contains TEs inserted upstream, downstream, or inside genes, which significantly alter gene expression levels across populations.  

```markdown
"Gene_name"	"NoInsertion (mean_TPM)"	"Upstream (mean_TPM)"	"Downstream (mean_TPM)"	"Inside (mean_TPM)"	"Upstream_vs_NoInsertion_logFoldChange"	"Inside_vs_NoInsertion_logFoldChange"	"Downstream_vs_NoInsertion_logFoldChange"
"AT4G34410"	562.145	NA	1.14	NA	NA	NA	-8.03975183205617
"AT3G48360"	150.99	0.16	16.46	NA	-7.033707790821	NA	-3.12185094231913
"AT4G01950"	389.61	2.33	17.3366666666667	NA	-6.87406289658701	NA	-4.4129255767341
"AT5G45890"	137.77	0.21	NA	NA	-6.84154485479315	NA	NA
"AT4G22880"	68.44	NA	0.07	NA	NA	NA	-6.02008424604333
...
```

Column Descriptions:
  1. `Gene_name`: Gene name.
  2. `NoInsertion (mean_TPM)`: Mean expression (TPM) of the gene with no TE insertion.
  3. `Upstream (mean_TPM)`: Mean expression (TPM) of the gene with a TE inserted upstream.
  4. `Downstream (mean_TPM)`: Mean expression (TPM) of the gene with a TE inserted downstream.
  5. `Inside (mean_TPM)`: Mean expression (TPM) of the gene with a TE inserted inside.
  6. `Upstream_vs_NoInsertion_logFoldChange`: Log2 fold change of upstream insertion compared to no insertion.
  7. `Inside_vs_NoInsertion_logFoldChange`: Log2 fold change of inside insertion compared to no insertion.
  8. `Downstream_vs_NoInsertion_logFoldChange`: Log2 fold change of downstream insertion compared to no insertion.  
Positive fold changes indicate upregulation, while negative values indicate downregulation. The formula for fold change is:  
`Upstream_vs_NoInsertion_logFoldChange = log2(Upstream + 1) - log2(NoInsertion + 1)`.


- **all_gene_TEs_details.tsv**  
This file details the positional relationships between TEs and genes across samples (e.g., upstream, downstream, inside, and distance), along with gene expression data.  

```markdown
"Gene_name"	"Genome_name"	"TE_name"	"Chromosome"	"TE_start"	"TE_end"	"Gene_start"	"Gene_end"	"Position"	"Species"	"Distance"	"expression"
"AT1G01020"	"44.ket_10.fa"	"44-Helitron_44"	"Chr1"	19241	20919	8958	11367	"Upstream"	"44.ket_10"	7874	12.65
"AT1G01030"	"44.ket_10.fa"	"44-Helitron_44"	"Chr1"	19241	20919	13881	15947	"Upstream"	"44.ket_10"	3294	0.98
"AT1G01040"	"44.ket_10.fa"	"44-Helitron_44"	"Chr1"	19241	20919	25358	33453	"Upstream"	"44.ket_10"	4439	31.61
"AT1G01050"	"44.ket_10.fa"	"25-TIR_69"	"Chr1"	44534	44768	33396	35397	"Upstream"	"44.ket_10"	9137	91.79
"AT1G01050"	"02.tibet.fa"	"25-TIR_69"	"Chr1"	47261	47494	36147	38160	"Upstream"	"02.tibet"	9101	102.58
"AT1G01050"	"25.per_1.fa"	"25-TIR_69"	"Chr1"	44736	44971	33603	35605	"Upstream"	"25.per_1"	9131	88.57
...
```

Column Descriptions:
  1. `Gene_name`: Gene name.
  2. `Genome_name`: Genome name.
  3. `TE_name`: TE name.
  4. `Chromosome`: Chromosome name.
  5. `TE_start` / `TE_end`: Start and end positions of the TE.
  6. `Gene_start` / `Gene_end`: Start and end positions of the gene.
  7. `Position`: TE position relative to the gene (e.g., upstream, downstream, inside).
  8. `Species`: Species name.
  9. `Distance`: Distance between the TE and the gene.
  10. `expression`: Gene expression level.

- **panHiTE.CorePan_fitmodel.pdf** and **panHiTE.CorePan_fitsmooth.pdf**  
Saturation curves of Pan and Core TE family counts based on a fitted model and LOESS smoothing.

<div style="text-align: center;">
    <img src="https://github.com/user-attachments/assets/96bb150b-e892-4db7-8c7d-3e418b3e8a29" alt="TE Family Proportion" width="400"/>
</div>

- **panTE.fa**  
A pan-TE library generated by panHiTE.

- **TE_summary.pdf**  
Statistical summaries of core, softcore, dispensable, and private TEs in the pan-genome, including:  

  - **Proportions of TE families** based on full-length TE annotations:  

  <div style="text-align: center;">
    <img src="https://github.com/user-attachments/assets/bd4361de-f240-4c0a-9a69-fbff5569063c" alt="TE Family Proportion" width="400"/>
  </div>

  - **Proportions of TE families** based on all TE annotations:  

  <div style="text-align: center;">
    <img src="https://github.com/user-attachments/assets/b58e41d5-4ffe-49f1-83af-55f3778ac5d8" alt="TE Family Proportion" width="400"/>
  </div>

  - **Genome coverage statistics** based on full-length TE annotations:  

  <div style="text-align: center;">
    <img src="https://github.com/user-attachments/assets/b1f7316f-eb6d-4cf4-b623-17aac62b9c36" alt="TE Family Proportion" width="400"/>
  </div>

  - **Genome coverage statistics** based on all TE annotations:  

  <div style="text-align: center;">
    <img src="https://github.com/user-attachments/assets/fe944d03-6540-4afe-a211-223116addf40" alt="TE Family Proportion" width="400"/>
  </div>

  - **Counts of different TE superfamilies** based on full-length TE annotations:  

  <div style="text-align: center;">
    <img src="https://github.com/user-attachments/assets/8b642309-6291-4ca0-bf64-8e0f74c82e55" alt="TE Family Proportion" width="400"/>
  </div>

  - **Counts of different TE superfamilies** based on all TE annotations:  
    
  <div style="text-align: center;">
    <img src="https://github.com/user-attachments/assets/91f163e6-64f7-4cb5-b998-696ed941a10c" alt="TE Family Proportion" width="400"/>
  </div>

  - **Genome coverage by TE superfamilies** based on full-length TE annotations:  

  <div style="text-align: center;">
    <img src="https://github.com/user-attachments/assets/47c653f9-1272-4d8e-a232-1d57352434e9" alt="TE Family Proportion" width="400"/>
  </div>

  - **Genome coverage by TE superfamilies** based on all TE annotations:
  
  <div style="text-align: center;">
    <img src="https://github.com/user-attachments/assets/193229c7-a8cf-4c83-9708-f800a819e4c6" alt="TE Family Proportion" width="400"/>
  </div>

  - **Insertion time distributions of Copia and Gypsy TEs** across samples (default mutation rate: 1.3e-8):  

  <div style="text-align: center;">
    <img src="https://github.com/user-attachments/assets/493ec975-2caf-4662-8f1e-bea6cb2f2b2d" alt="TE Family Proportion" width="400"/>
  </div>

---

## panHiTE Tutorial

In this tutorial, we'll demonstrate how to use panHiTE with a demo data.

### 1. Installation via Conda  
*(Docker and Singularity versions will be available once panHiTE stabilizes.)*

```bash
git clone https://github.com/CSU-KangHu/HiTE.git

cd HiTE && python configure.py

source ~/.bashrc # or open a new terminal

conda env create --name HiTE -f environment.yml
conda activate HiTE
```

### 2. Installing Required R Packages
To install R packages, you may need to configure the CRAN mirror first. For example, to use the default CRAN mirror, add the following configuration to your `~/.Rprofile` file:

```R
options(repos = c(CRAN = "https://cloud.r-project.org/"))
```

We provide a script to install all necessary R dependencies in one step:  
```bash
conda activate HiTE
Rscript RNA_seq/install_R_dependencies.R
```

If the installation completes successfully, the following R packages will be installed within the `HiTE` environment:  
- **argparser**  
- **tibble**  
- **dplyr**  
- **minpack.lm**  
- **readr**  
- **stringr**  
- **tidyr**  
- **Rsubread**
- **limma** 
- **edgeR**
- **ggplot2**

If the script fails to install the packages, you can also install them manually.

---

### 3. Data Preparation  

Download the [demo data](https://zenodo.org/records/14235586) from Zenodo. Since the dataset includes RNA-seq data, it is relatively large, so please be patient.

1. **pan_genomes_dir (Required)**  
   All genome assemblies should be stored in a single directory, which should be specified as the `pan_genomes_dir` parameter.

2. **genome_list (Required)**  
   A tab-delimited file with the following columns:  
   - Column 1: Genome assembly file name  
   - Column 2: Gene annotation file name (optional)  
   - Columns 3 & 4: Paths to RNA-seq data (optional, single-end data in column 3, paired-end data in columns 3 & 4)  

   Example:  
   ```markdown
    44.ket_10.fa    44.ket_10.gff   CRR624282_Chr1_f1.fq.gz CRR624282_Chr1_r2.fq.gz
    02.tibet.fa     02.tibet.gff    CRR624279_Chr1_f1.fq.gz CRR624279_Chr1_r2.fq.gz
    25.per_1.fa     25.per_1.gff    SRR748686_Chr1.fq.gz
    ```

3. **out_dir (Required)**  
   Specify the output directory path.

4. **gene_dir (Optional)**  
   Place all gene annotation files in a single directory and set this directory as the `gene_dir` parameter.  
   **Important**: Ensure that the gene_id in multiple gene annotation files has a **consistent name**, with the last element separated by an `underscore`.  
   For example, in the file 01.col.gtf, a gene_id might be `col_AT1G01010`, and in the file 02.tibet.gtf, the gene_id should be `tibet_AT1G01010`.

5. **RNA_dir (Optional)**  
   Set this as the parent directory for RNA-seq reads. The paths listed in columns 3 & 4 of the `genome_list` should be accessible via this directory.

---

### 4. Running panHiTE  

#### 4.1 Full Workflow  
To run panHiTE from start to end, use the following command:  
```bash
# Make sure to replace the `${}` placeholders with your actual paths.
python panHiTE.py \
 --pan_genomes_dir ${pan_genomes_dir} \
 --genome_list ${genome_list} \
 --genes_dir ${gene_dir} \
 --RNA_dir ${RNA_dir} \
 --out_dir ${out_dir} \
 --thread ${threads} \
 --recover 1 \
 --skip_analyze 0
 
# Example script:
#source_dir=/public/home/xxx/test/panHiTE
#pan_genomes_dir=/public/home/xxx/ath_pan_genome/pan_genome/ath/ath_genome_and_annotation/genomes
#gene_dir=/public/home/xxx/ath_pan_genome/pan_genome/ath/ath_genome_and_annotation/gtf_files
#RNA_dir=/public/home/xxx/ath_pan_genome/pan_genome/ath/ath_genome_and_annotation/RNA_seq_files
#genome_list=/public/home/xxx/ath_pan_genome/pan_genome/ath/ath_genome_and_annotation/genome_list
#out_dir=/public/home/xxx/ath_pan_genome/pan_genome/ath/panHiTE_serial_output
#cd $source_dir && /usr/bin/time -v python panHiTE.py \
# --pan_genomes_dir ${pan_genomes_dir} \
# --genome_list ${genome_list} \
# --genes_dir ${gene_dir} \
# --RNA_dir ${RNA_dir} \
# --out_dir ${out_dir} \
# --thread 40 \
# --recover 1 \
# --te_type all \
# --skip_analyze 0 \
# --miu 7e-9
```
- **`--recover 1`**: This option detects existing intermediate files and resumes from the last checkpoint, saving time.

#### 4.2 skip downstream analysis 
If you only need the panLTR library and the annotation for each genome, you can choose to skip the downstream analysis by setting `--skip_analyze 1`.
```bash
python panHiTE.py \
 --pan_genomes_dir ${pan_genomes_dir} \
 --genome_list ${genome_list} \
 --out_dir ${out_dir} \
 --thread ${threads} \
 --recover 1 \
 --te_type all \
 --skip_analyze 1 \
 --miu ${miu}

# Example script:
#source_dir=/public/home/xxx/test/panHiTE
#pan_genomes_dir=/public/home/xxx/ath_pan_genome/pan_genome/ath/ath_genome_and_annotation/genomes
#gene_dir=/public/home/xxx/ath_pan_genome/pan_genome/ath/ath_genome_and_annotation/gtf_files
#RNA_dir=/public/home/xxx/ath_pan_genome/pan_genome/ath/ath_genome_and_annotation/RNA_seq_files
#genome_list=/public/home/xxx/ath_pan_genome/pan_genome/ath/ath_genome_and_annotation/genome_list
#out_dir=/public/home/xxx/ath_pan_genome/pan_genome/ath/panHiTE_serial_output
#cd $source_dir && /usr/bin/time -v python panHiTE.py \
# --pan_genomes_dir ${pan_genomes_dir} \
# --genome_list ${genome_list} \
# --out_dir ${out_dir} \
# --thread 40 \
# --recover 1 \
# --te_type all \
# --skip_analyze 1 \
# --miu 7e-9
```
---

### 5. Features Under Development

The current panHiTE pipeline runs HiTE detection on multiple genomes sequentially, generates the panHiTE library, 
and then annotates each genome sequentially as well. In theory, these steps can be parallelized on an HPC platform. 
Therefore, we plan to develop an HPC-compatible panHiTE version based on Nextflow, enabling faster and more efficient panHiTE analysis.
